Reduction on expanded tensor produces result with incorrect shape

Here's an example of two almost identical fusions with the only difference being the contiguity of T0, the output of the fusion with `contiguity=[None, None]` produces the result with incorrect shape:
```py
from nvfuser import FusionDefinition, DataType
import torch

# The only difference between two fusions is "contiguity" of T0.
def nvfuser_fusion_id0(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1], contiguity=[None, None], dtype=DataType.Float, is_cpu=False, stride_order=[1, 0])
    T1 = fd.ops.sum(T0, axes=[0], keepdim=False, dtype=DataType.Null)
    fd.add_output(T1)

def nvfuser_fusion_id1(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[-1, -1], contiguity=[True, True], dtype=DataType.Float, is_cpu=False, stride_order=[1, 0])
    T1 = fd.ops.sum(T0, axes=[0], keepdim=False, dtype=DataType.Null)
    fd.add_output(T1)

t0_contiguous = torch.ones((2, 3), dtype=torch.float32, device='cuda')
t0_expanded = torch.tensor(1.0, device='cuda').as_strided((2, 3), (0, 0))

with FusionDefinition() as fd0:
    nvfuser_fusion_id0(fd0)

with FusionDefinition() as fd1:
    nvfuser_fusion_id1(fd1)

out0 = fd0.execute([t0_expanded])
out1 = fd1.execute([t0_contiguous])

# AssertionError: The values for attribute 'shape' do not match: torch.Size([1]) != torch.Size([3]).
torch.testing.assert_allclose(out0[0], out1[0])
```

--- a/csrc/ir/nodes.cpp
+++ b/csrc/ir/nodes.cpp
@@ -1309,8 +1309,6 @@ SqueezeOp::SqueezeOp(
           id->isBroadcast() || id->isSymbolic(),
           "Squeeze dimension should be either Symbolic or Broadcast. Found ",
           id->getIterType());
-      NVF_ERROR(
-          !id->hasExpandedExtent(), "Can not squeeze expanded dimension(s).");
       if (id->isBroadcast()) {
         // Check concrete broadcast extent here. For Symbolic inputs, this check
         // will be deferred to concretization. See dynamic_transform.cpp

--- a/csrc/ir/nodes.cpp
+++ b/csrc/ir/nodes.cpp
@@ -2101,7 +2101,7 @@ std::vector<PolymorphicValue> ExpandOp::evaluate(
   for (auto i : c10::irange(1, inputs.size())) {
     expanded_size.push_back((int64_t)inputs.at(i));
   }
-  return {at::expand_copy(in, expanded_size)};
+  return {in.expand(expanded_size)};
 }
 
 NVFUSER_DEFINE_CLONE_AND_CREATE(ExpandOp)

--- a/csrc/ir/nodes.cpp
+++ b/csrc/ir/nodes.cpp
@@ -1354,12 +1354,20 @@ std::vector<PolymorphicValue> SqueezeOp::evaluate(
   NVF_ERROR(
       (int64_t)is_squeeze_dims.size() == in.dim(),
       "The dimensions of input tensor and does not match with is_squeeze_dims");
+  at::Tensor out = in;
   for (int64_t i : c10::irange((int64_t)is_squeeze_dims.size())) {
-    if (!is_squeeze_dims[i]) {
+    if (is_squeeze_dims[i]) {
+      if (in.stride(i) == 0) {
+        // If the input dimension is expanded in this dimension, undo the expand
+        // by slicing. This ensures that any broadcast dimensions will be
+        // unexpanded when we do the final call to view()
+        out = out.slice(i, 0, 1);
+      }
+    } else {
       out_shape.push_back(in.sizes()[i]);
     }
   }
-  return {in.view(out_shape)};
+  return {out.view(out_shape)};
 }
 
 void SqueezeOp::checkConcretization(Val* old_val, Val* new_val) const {

--- a/csrc/ops/alias.cpp
+++ b/csrc/ops/alias.cpp
@@ -245,8 +245,6 @@ TensorView* squeeze(TensorView* x, const std::vector<int64_t>& dims) {
           continue;
         }
         NVF_CHECK(
-            !id->hasExpandedExtent(), "Can not squeeze expanded dimension(s).");
-        NVF_CHECK(
             id->extent()->isConstScalar() && id->extent()->evaluate() == 1,
             "Can not squeeze dimension(s) with size != 1.");
       }
@@ -291,8 +289,6 @@ TensorView* squeeze(TensorView* x, const std::vector<bool>& to_squeeze) {
         NVF_CHECK(
             id->isBroadcast(),
             "Can not squeeze non-broadcasting dimension(s).");
-        NVF_CHECK(
-            !id->hasExpandedExtent(), "Can not squeeze expanded dimension(s).");
         NVF_CHECK(
             id->extent()->isConstScalar() && id->extent()->evaluate() == 1,
             "Can not squeeze dimension(s) with size != 1.");

--- a/csrc/ops/alias.cpp
+++ b/csrc/ops/alias.cpp
@@ -204,7 +204,10 @@ TensorView* flatten(TensorView* x, int64_t start_dim, int64_t end_dim) {
   return out;
 }
 
-TensorView* squeeze(TensorView* x, const std::vector<int64_t>& dims) {
+TensorView* squeeze(
+    TensorView* x,
+    const std::vector<int64_t>& dims,
+    bool squeeze_expanded) {
   NVF_ERROR(x != nullptr, "Input is invalid.");
   auto x_dom = x->domain()->noReductions();
   const auto ndims = static_cast<int>(x_dom.size());
@@ -232,44 +235,13 @@ TensorView* squeeze(TensorView* x, const std::vector<int64_t>& dims) {
     to_squeeze[dim] = true;
   }
 
-  std::vector<IterDomain*> out_domain;
-  for (const auto idx : c10::irange(ndims)) {
-    auto id = x_dom[idx];
-    if (to_squeeze[idx]) {
-      if (!id->isSymbolic()) {
-        // If a squeeze is attempted on a non-broadcast dimension
-        // just don't do it!  This conforms with Pytorch.
-        if (!id->isBroadcast()) {
-          to_squeeze[idx] = false;
-          out_domain.push_back(id->cloneWithoutRFactor());
-          continue;
-        }
-        NVF_CHECK(
-            id->extent()->isConstScalar() && id->extent()->evaluate() == 1,
-            "Can not squeeze dimension(s) with size != 1.");
-      }
-    } else {
-      out_domain.push_back(id->cloneWithoutRFactor());
-    }
-  }
-
-  auto out = IrBuilder::create<TensorView>(
-      IrBuilder::create<TensorDomain>(
-          out_domain, TensorDomain::getContiguityFilledWith(out_domain, true)),
-      *x->getDataType());
-
-  std::vector<bool> all_false(to_squeeze.size(), false);
-  // If a squeeze does not perform a squeeze, create a no-op
-  if (to_squeeze == all_false) {
-    IrBuilder::create<LoadStoreOp>(LoadStoreOpType::Set, out, x);
-  } else {
-    IrBuilder::create<SqueezeOp>(x->container(), out, x, to_squeeze);
-  }
-
-  return out;
+  return squeeze(x, to_squeeze, squeeze_expanded);
 }
 
-TensorView* squeeze(TensorView* x, const std::vector<bool>& to_squeeze) {
+TensorView* squeeze(
+    TensorView* x,
+    const std::vector<bool>& to_squeeze,
+    bool squeeze_expanded) {
   NVF_ERROR(x != nullptr, "Input is invalid.");
   auto x_dom = x->domain()->noReductions();
   const auto ndims = static_cast<int>(x_dom.size());
@@ -290,6 +262,11 @@ TensorView* squeeze(TensorView* x, const std::vector<bool>& to_squeeze) {
             id->isBroadcast(),
             "Can not squeeze non-broadcasting dimension(s).");
         NVF_CHECK(
+            squeeze_expanded || !id->hasExpandedExtent(),
+            "Refusing to squeeze expanded IterDomain ",
+            id->toString(),
+            ". To force removal of this axis, use squeeze_expanded=true.");
+        NVF_CHECK(
             id->extent()->isConstScalar() && id->extent()->evaluate() == 1,
             "Can not squeeze dimension(s) with size != 1.");
       }
@@ -303,7 +280,13 @@ TensorView* squeeze(TensorView* x, const std::vector<bool>& to_squeeze) {
           out_domain, TensorDomain::getContiguityFilledWith(out_domain, true)),
       *x->getDataType());
 
-  IrBuilder::create<SqueezeOp>(x->container(), out, x, to_squeeze);
+  if (std::none_of(
+          to_squeeze.begin(), to_squeeze.end(), [](bool b) { return b; })) {
+    // If we did not squeeze any axes, this is just set()
+    IrBuilder::create<LoadStoreOp>(LoadStoreOpType::Set, out, x);
+  } else {
+    IrBuilder::create<SqueezeOp>(x->container(), out, x, to_squeeze);
+  }
 
   return out;
 }

--- a/csrc/ops/alias.h
+++ b/csrc/ops/alias.h
@@ -43,12 +43,26 @@ TensorView* reshape(TensorView* x, const std::vector<Val*>& new_sizes);
 
 TensorView* flatten(TensorView* x, int64_t start_dim = 0, int64_t end_dim = -1);
 
-// This implementation is specific to Pytorch where if you attempt to squeeze
-// a non-broadcast dimension, the squeeze does not do anything to that
-// dimension and does not trigger an error.
-TensorView* squeeze(TensorView* x, const std::vector<int64_t>& dims);
+//! This implementation differs from Pytorch where if you attempt to squeeze
+//! a non-broadcast dimension, the squeeze does not do anything to that
+//! dimension and does not trigger an error. Here, we will raise an exception in
+//! such cases. It is also an error to request a squeeze of an expanded
+//! dimension; however, if squeeze_expanded is true then this is allowed and
+//! works just like if the dimension was not expanded.
+TensorView* squeeze(
+    TensorView* x,
+    const std::vector<int64_t>& dims,
+    bool squeeze_expanded = false);
 
-TensorView* squeeze(TensorView* x, const std::vector<bool>& to_squeeze);
+//! Squeeze the dimensions corresponding to "true" in to_squeeze, i.e. remove
+//! those broadcasted dimension. If squeeze_expanded is true, then expanded
+//! Broadcasts will be removed just as if they were not expanded. If
+//! squeeze_expanded is false, then it is an error for an expanded broadcast to
+//! have a corresponding "true" value in to_squeeze.
+TensorView* squeeze(
+    TensorView* x,
+    const std::vector<bool>& to_squeeze,
+    bool squeeze_expanded = false);
 
 TensorView* unsqueeze(TensorView* x, int dim);
 

--- a/csrc/ops/arith.cpp
+++ b/csrc/ops/arith.cpp
@@ -1299,17 +1299,20 @@ TensorView* reductionOp(
   //   Add -> multiplication by i0
   //   Mul -> raise to power i0
   //   Min/Max -> squeeze
-  //   {Logical,Bitwise}{And,Or,Xor} -> squeeze
+  //   {Logical,Bitwise}{And,Or} -> squeeze
+  //   BitwiseXor -> 0 if i0 is even, else squeeze
   //   Eq -> squeeze
   //   Gcd -> squeeze
   // Other op-types are non-commutative, so we ignore them here as they should
-  // not be used in reductions. We can see that the only two that require
-  // special consideration are Add and Mul. We treat all others as trivial (i.e.
+  // not be used in reductions. We can see that the only two common ops that
+  // require special consideration are Add and Mul. Currently Xor is not
+  // supported for expanded reduction. We treat all others as trivial (i.e.
   // squeeze).
   std::vector<int> reduction_axes;
   std::vector<bool> is_broadcast_reduction(ndims, false);
   bool expand_reductions_are_trivial = reduction_op_type != BinaryOpType::Add &&
-      reduction_op_type != BinaryOpType::Mul;
+      reduction_op_type != BinaryOpType::Mul &&
+      reduction_op_type != BinaryOpType::BitwiseXor;
   int offset = 0;
   for (unsigned int axis : uint_axes) {
     auto id = tv_root[axis];

--- a/csrc/ops/arith.cpp
+++ b/csrc/ops/arith.cpp
@@ -1309,7 +1309,7 @@ TensorView* reductionOp(
   // supported for expanded reduction. We treat all others as trivial (i.e.
   // squeeze).
   std::vector<int> reduction_axes;
-  std::vector<bool> is_broadcast_reduction(ndims, false);
+  std::vector<bool> is_squeeze(ndims, false);
   bool expand_reductions_are_trivial = reduction_op_type != BinaryOpType::Add &&
       reduction_op_type != BinaryOpType::Mul &&
       reduction_op_type != BinaryOpType::BitwiseXor;
@@ -1317,7 +1317,7 @@ TensorView* reductionOp(
   for (unsigned int axis : uint_axes) {
     auto id = tv_root[axis];
     if (id->isBroadcast()) {
-      is_broadcast_reduction[axis] = true;
+      is_squeeze[axis] = true;
       offset--;
     } else {
       reduction_axes.push_back((int)axis + offset);
@@ -1327,7 +1327,7 @@ TensorView* reductionOp(
   TensorView* squeezed = tv;
   if (offset < 0) {
     // There are some broadcast dims being reduced. We squeeze them all first.
-    squeezed = squeeze(tv, is_broadcast_reduction, /*squeeze_expanded=*/true);
+    squeezed = squeeze(tv, is_squeeze, /*squeeze_expanded=*/true);
   }
 
   TensorView* out = squeezed;

--- a/csrc/ops/arith.cpp
+++ b/csrc/ops/arith.cpp
@@ -1327,7 +1327,7 @@ TensorView* reductionOp(
   TensorView* squeezed = tv;
   if (offset < 0) {
     // There are some broadcast dims being reduced. We squeeze them all first.
-    squeezed = squeeze(tv, is_broadcast_reduction);
+    squeezed = squeeze(tv, is_broadcast_reduction, /*squeeze_expanded=*/true);
   }
 
   TensorView* out = squeezed;
@@ -1809,7 +1809,7 @@ WelfordResult Welford(
 
   TensorView* squeezed = tv;
   if (offset < 0) {
-    squeezed = squeeze(tv, is_trivial_reduction);
+    squeezed = squeeze(tv, is_trivial_reduction, /*squeeze_expanded=*/true);
   }
 
   if (!reduction_axes.empty()) {

--- a/csrc/ops/arith.cpp
+++ b/csrc/ops/arith.cpp
@@ -1291,30 +1291,69 @@ TensorView* reductionOp(
     return maybe_full;
   }
 
+  // [Trivial reductions]
+  // When we reduce a simple broadcast axis like bS0{1} the effect is just to
+  // squeeze out that broadcast axis. When the axis is expanded, such as bS1{1
+  // ex i0}, then the effect depends on the op type. We have the following
+  // mappings from op_type to expanded reduction equivalent:
+  //   Add -> multiplication by i0
+  //   Mul -> raise to power i0
+  //   Min/Max -> squeeze
+  //   {Logical,Bitwise}{And,Or,Xor} -> squeeze
+  //   Eq -> squeeze
+  //   Gcd -> squeeze
+  // Other op-types are non-commutative, so we ignore them here as they should
+  // not be used in reductions. We can see that the only two that require
+  // special consideration are Add and Mul. We treat all others as trivial (i.e.
+  // squeeze).
   std::vector<int> reduction_axes;
-  std::vector<bool> is_trivial_reduction(ndims, false);
+  std::vector<bool> is_broadcast_reduction(ndims, false);
+  bool expand_reductions_are_trivial = reduction_op_type != BinaryOpType::Add &&
+      reduction_op_type != BinaryOpType::Mul;
   int offset = 0;
   for (unsigned int axis : uint_axes) {
     auto id = tv_root[axis];
-    is_trivial_reduction[axis] = id->isBroadcast() &&
-        !id->hasExpandedExtent() && id->extent()->isConstInt() &&
-        id->extent()->evaluate().as<int64_t>() == 1;
-    if (!is_trivial_reduction[axis]) {
-      reduction_axes.push_back((int)axis + offset);
-    } else if (!keep_dim) {
+    if (id->isBroadcast()) {
+      is_broadcast_reduction[axis] = true;
       offset--;
+    } else {
+      reduction_axes.push_back((int)axis + offset);
     }
   }
 
   TensorView* squeezed = tv;
   if (offset < 0) {
-    squeezed = squeeze(tv, is_trivial_reduction);
+    // There are some broadcast dims being reduced. We squeeze them all first.
+    squeezed = squeeze(tv, is_broadcast_reduction);
   }
 
   TensorView* out = squeezed;
   if (!reduction_axes.empty()) {
-    return reductionOpRaw(
+    out = reductionOpRaw(
         reduction_op_type, reduction_axes, init, squeezed, keep_dim, dtype);
+  }
+
+  if (!expand_reductions_are_trivial) {
+    Val* factor = nullptr;
+    for (auto axis : uint_axes) {
+      IterDomain* id = tv_root[axis];
+      if (id->isBroadcast() && id->hasExpandedExtent()) {
+        factor =
+            SimplifyingIrBuilder::mulExpr(factor, id->getMaybeExpandedExtent());
+      }
+    }
+    if (factor != nullptr) {
+      factor = SimplifyingIrBuilder::maybeCastExpr(out->dtype(), factor);
+      if (reduction_op_type == BinaryOpType::Add) {
+        out = mul(out, factor);
+      } else if (reduction_op_type == BinaryOpType::Mul) {
+        out = pow(out, factor);
+      } else {
+        NVF_ERROR(
+            false,
+            "Add and Mul are the only non-trivial expand reductions allowed");
+      }
+    }
   }
 
   if (out == tv) {

--- a/python_tests/test_python_frontend.py
+++ b/python_tests/test_python_frontend.py
@@ -3151,6 +3151,26 @@ class TestNvFuserFrontend(TestCase):
         nvf_out, _ = self.exec_nvfuser(fusion_func, inputs)
         # self.assertEqual(nvf_out[0], t24)
 
+    # Test that expanded dimensions can be reduced properly
+    # See https://github.com/NVIDIA/Fuser/issues/1678
+    def test_expanded_reduction(self):
+        inputs = [torch.tensor(1.0, device="cuda").as_strided((2, 3), (0, 0))]
+
+        def fusion_func(fd: FusionDefinition) -> None:
+            T0 = fd.define_tensor(
+                shape=[-1, -1],
+                contiguity=[None, None],
+                dtype=DataType.Float,
+                is_cpu=False,
+                stride_order=[1, 0],
+            )
+            T1 = fd.ops.sum(T0, axes=[0], keepdim=False, dtype=DataType.Null)
+            fd.add_output(T1)
+
+        nvf_out, _ = self.exec_nvfuser(fusion_func, inputs)
+
+        self.assertEqual(nvf_out[0], inputs[0].sum(dim=0))
+
 
 if __name__ == "__main__":
     run_tests()

--- a/test/test_gpu3.cpp
+++ b/test/test_gpu3.cpp
@@ -3496,58 +3496,6 @@ TEST_F(NVFuserTest, FusionExpandReduce_CUDA) {
   testValidate(executor_cache.fusion(), cg_outputs, {t0}, __LINE__, __FILE__);
 }
 
-// Predicate elimination issue repro:
-TEST_F(NVFuserTest, FusionExpandReduce2_CUDA) {
-  auto fusion = std::make_unique<Fusion>();
-  FusionGuard fg(fusion.get());
-
-  auto tv0 = makeConcreteTensor({1, 4});
-  fusion->addInput(tv0);
-
-  auto tv1 =
-      expand(tv0, {IrBuilder::create<Val>(3L), IrBuilder::create<Val>(4L)});
-
-  auto tv2 = sum(tv1, {0});
-  fusion->addOutput(tv2);
-
-  // tv2[r{3}, i{4}]
-  tv2->split(0, NamedScalar::getParallelDim(ParallelType::TIDy));
-  tv2->axis(1)->parallelize(ParallelType::TIDy);
-  tv2->split(0, NamedScalar::getParallelDim(ParallelType::BIDy), false);
-  tv2->axis(0)->parallelize(ParallelType::BIDy);
-  tv2->split(-1, NamedScalar::getParallelDim(ParallelType::TIDx));
-  tv2->axis(-1)->parallelize(ParallelType::TIDx);
-  tv2->axis(-2)->parallelize(ParallelType::BIDx);
-  // [rBIDy, rO, rTIDy, iBIDx, iTIDx]
-  tv2->reorder({{-2, 0}, {-1, 1}, {2, 2}});
-  // [iBIDx, iTIDx, rTIDy, rBIDy, rO]
-  auto tv3 = tv2->rFactor({-1});
-
-  TransformPropagatorWithCheck propagator(tv3);
-  MaxRootDomainInfoSpanningTree(tv3).traverse(&propagator);
-  scheduler_utils::parallelizeAllLike(tv3);
-  tv0->computeAt(tv3, -1, ComputeAtMode::MostInlined);
-
-  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
-  auto t0 = at::randn({1, 4}, options);
-
-  FusionExecutor fe;
-  fe.compileFusion(fusion.get(), {t0}, LaunchParams(-1, 2, -1, 4, 2, 1));
-  auto cg_outputs = fe.runFusion({t0}, LaunchParams(-1, 2, -1, 4, 2, 1));
-
-  auto ref = t0.expand({3, 4}).sum({0});
-
-  testValidate(
-      fusion.get(),
-      cg_outputs,
-      {t0},
-      {ref},
-      __LINE__,
-      __FILE__,
-      "",
-      LaunchParams(-1, 2, -1, 4, 2, 1));
-}
-
 TEST_F(NVFuserTest, FusionVectorComponentReduce_CUDA) {
   auto fusion = std::make_unique<Fusion>();
   FusionGuard fg(fusion.get());

